lib/db_utils.go

```diff
@@ -10661,6 +10681,409 @@ func DBDeletePostAssociationWithTxn(txn *badger.Txn, snap *Snapshot, association
+	DecodeFromBytes(lockupYieldCurvePointObj, rr)
+	return lockupYieldCurvePointObj
+}
+
+func DBGetLocalYieldCurvePoints(handle *badger.DB, snap *Snapshot, profilePKID *PKID, lockupDurationNanoSecs int64) (
+	_leftLockupYieldCurvePoint *LockupYieldCurvePoint, _rightLockupYieldCurvePoint *LockupYieldCurvePoint) {
+	var leftLockupYieldCurvePoint *LockupYieldCurvePoint
+	handle.View(func(txn *badger.Txn) error {
+		leftLockupYieldCurvePoint = DBGetLeftLockupYieldCurvePointWithTxn(
+			txn, snap, profilePKID, lockupDurationNanoSecs)
+		return nil
+	})
+	var rightLockupYieldCurvePoint *LockupYieldCurvePoint
+	handle.View(func(txn *badger.Txn) error {
+		rightLockupYieldCurvePoint = DBGetRightLockupYieldCurvePointWithTxn(
+			txn, snap, profilePKID, lockupDurationNanoSecs)
+		return nil
+	})
+	return leftLockupYieldCurvePoint, rightLockupYieldCurvePoint
+}
+
+func DBGetLeftLockupYieldCurvePointWithTxn(txn *badger.Txn, snap *Snapshot, profilePKID *PKID,
+	lockupDurationNanoSecs int64) (_leftLockupYieldCurvePoint *LockupYieldCurvePoint) {
+	key := _dbKeyForLockupYieldCurvePoint(LockupYieldCurvePoint{
+		ProfilePKID:            profilePKID,
+		LockupDurationNanoSecs: lockupDurationNanoSecs,
+	})
+
+	// Seek left of the yield curve point.
+	iterLeftOpts := badger.DefaultIteratorOptions
+	iterLeftOpts.Reverse = true
+	iterLeft := txn.NewIterator(iterLeftOpts)
+	iterLeft.Seek(key)
+	iterLeftKey := iterLeft.Item().Key()
+
+	// There's a chance our seek yield a key in a different prefix (i.e. not a yield curve point).
+	// In this case, we know _dbKeyToLockupYieldCurvePoint will fail in parsing the key.
+	// We can return early in this case as there's no relevant yield points in the DB.
+	if len(iterLeftKey) < len(Prefixes.PrefixLockupYieldCurvePointByProfilePKIDAndDurationNanoSecs) {
+		return nil
+	}
+	if !bytes.Equal(iterLeftKey[:len(Prefixes.PrefixLockupYieldCurvePointByProfilePKIDAndDurationNanoSecs)],
+		Prefixes.PrefixLockupYieldCurvePointByProfilePKIDAndDurationNanoSecs) {
+		return nil
+	}
+
+	// Fetch the LockupYieldCurvePoint beneath the key.
+	leftLockupYieldCurvePointBytes, err := DBGetWithTxn(txn, snap, iterLeftKey)
+	if err != nil {
+		return nil
+	}
+
+	// Parse the bytes beneath the key.
+	leftLockupYieldCurvePointObj := &LockupYieldCurvePoint{}
+	rr := bytes.NewReader(leftLockupYieldCurvePointBytes)
+	DecodeFromBytes(leftLockupYieldCurvePointObj, rr)
+	return leftLockupYieldCurvePointObj
+}
+
+func DBGetRightLockupYieldCurvePointWithTxn(txn *badger.Txn, snap *Snapshot, profilePKID *PKID,
+	lockupDurationNanoSecs int64) (_rightLockupYieldCurvePoint *LockupYieldCurvePoint) {
+	key := _dbKeyForLockupYieldCurvePoint(LockupYieldCurvePoint{
+		ProfilePKID:            profilePKID,
+		LockupDurationNanoSecs: lockupDurationNanoSecs,
+	})
+
+	// Seek left of the yield curve point.
+	iterRightOpts := badger.DefaultIteratorOptions
+	iterRight := txn.NewIterator(iterRightOpts)
+	iterRight.Seek(key)
+	iterRightKey := iterRight.Item().Key()
```

Comment:
I'm not sure this does what we want. Badger always seeks to a key "less than or equal" to the one you're seeking. So if you have [1, 2, 3] and you seek for 2.5, for example, it will return 2. But the value that's right of 2.5 is actually 3.

I could be wrong but just noodle on it and make sure. Also check the left seeker to verify this as well. But I think the left seeker works properly by default because "less than or equal" does what we want.

Reply:
This is definitely going to be tested, but I think it's fine per the Badger documentation here: https://pkg.go.dev/github.com/dgraph-io/badger#section-readme

It states, "Just like forward iteration goes to the first key which is equal or greater than the SEEK key, reverse iteration goes to the first key which is equal or lesser than the SEEK key. Therefore, SEEK key would not be part of the results."

I'll follow up on this.